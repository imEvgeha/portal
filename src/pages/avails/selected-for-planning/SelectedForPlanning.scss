.selected-for-planning-table-wrapper {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    height: 100%;
    min-height: 400px;
    padding: 24px 24px 0;
    overflow: auto;
}
