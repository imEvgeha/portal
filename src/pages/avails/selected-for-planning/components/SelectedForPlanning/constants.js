export const EXCLUDED_ATTRIBUTES = ['PROJECT_NAME'];
export const DOP_PROJECT_STATUS_COMPLETED = 'COMPLETED';
export const DOP_PROJECT_STATUS_IN_PROGRESS = 'CIN PROGRESS';
export const DOP_PROJECT_STATUS_NOT_STARTED = 'BNOT STARTED';
export const DOP_PROJECT_STATUS_CANCELLED = 'CANCELLED';
export const EXCLUDED_STATUSES = [DOP_PROJECT_STATUS_CANCELLED, DOP_PROJECT_STATUS_COMPLETED];
