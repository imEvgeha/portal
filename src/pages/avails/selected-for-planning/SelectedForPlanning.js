import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {cloneDeep} from 'lodash';
import config from 'react-global-configuration';
import {compose} from 'redux';
import NexusGrid from '../../../ui/elements/nexus-grid/NexusGrid';
import {GRID_EVENTS} from '../../../ui/elements/nexus-grid/constants';
import withColumnsResizing from '../../../ui/elements/nexus-grid/hoc/withColumnsResizing';
import withFilterableColumns from '../../../ui/elements/nexus-grid/hoc/withFilterableColumns';
import withInfiniteScrolling from '../../../ui/elements/nexus-grid/hoc/withInfiniteScrolling';
import withSideBar from '../../../ui/elements/nexus-grid/hoc/withSideBar';
import withSorting from '../../../ui/elements/nexus-grid/hoc/withSorting';
import {prepareSelectForPlanningData} from './utils';
import {COLUMN_MAPPINGS, DOP_PROJECT_URL, SELECTED_FOR_PLANNING_TAB} from './constants';

const SelectedForPlanningTable = compose(
    withFilterableColumns(),
    withColumnsResizing(),
    withSideBar(),
    withSorting(),
    withInfiniteScrolling({fetchData: prepareSelectForPlanningData})
)(NexusGrid);

const SelectedForPlanning = ({
    activeTab,
    isPlanningTabRefreshed,
    setSelectedForPlanningGridApi,
    setSelectedForPlanningColumnApi,
}) => {
    const [updatedColDef, setUpdatedColDef] = useState([]);
    const [externalSort, setExternalSort] = useState({});

    const mappings = COLUMN_MAPPINGS.map(col =>
        col.colId === 'projectId'
            ? {
                  ...col,
                  cellRendererParams: {
                      link: `${config.get('gateway.DOPUrl')}${DOP_PROJECT_URL}`,
                  },
              }
            : col
    );

    const onGridReady = ({type, columnApi, api}) => {
        switch (type) {
            case GRID_EVENTS.READY: {
                setSelectedForPlanningColumnApi(columnApi);
                setSelectedForPlanningGridApi(api);
                break;
            }
            default:
            // no-op;
        }
    };

    const onSortChanged = ({api}) => {
        // prepare sorting as a payload
        const sortModel = api.getSortModel();
        if (sortModel.length) {
            const sortCriterion = [
                {
                    fieldName: sortModel[0].colId,
                    ascending: sortModel[0].sort === 'asc',
                },
            ];
            setExternalSort(prevData => {
                return {
                    ...prevData,
                    sortCriterion,
                };
            });
        } else {
            setExternalSort({});
        }
    };

    const dragStoppedHandler = event => {
        const updatedMappings = updatedColDef.length ? cloneDeep(updatedColDef) : cloneDeep(mappings);
        const columnHeader = event.target.textContent.trim();
        const columns = event.columnApi.columnController.gridColumns;

        const moveTo = columns.findIndex(col => col.colDef.headerName === columnHeader);
        const moveFrom = updatedMappings.findIndex(col => col.headerName === columnHeader);
        const [movedColumn] = updatedMappings.splice(moveFrom, 1);
        updatedMappings.splice(moveTo, 0, movedColumn);

        setUpdatedColDef(updatedMappings);
    };

    return (
        <SelectedForPlanningTable
            id="selectedForPlanningRepo"
            columnDefs={updatedColDef.length ? updatedColDef : mappings}
            mapping={COLUMN_MAPPINGS}
            rowSelection="multiple"
            suppressRowClickSelection
            isGridHidden={activeTab !== SELECTED_FOR_PLANNING_TAB}
            key={`planning_table_${isPlanningTabRefreshed}`}
            onGridEvent={onGridReady}
            dragStopped={dragStoppedHandler}
            onSortChanged={onSortChanged}
            externalFilter={externalSort}
        />
    );
};

SelectedForPlanning.propTypes = {
    activeTab: PropTypes.string.isRequired,
    isPlanningTabRefreshed: PropTypes.bool.isRequired,
    setSelectedForPlanningGridApi: PropTypes.func,
    setSelectedForPlanningColumnApi: PropTypes.func,
};

SelectedForPlanning.defaultProps = {
    setSelectedForPlanningGridApi: () => null,
    setSelectedForPlanningColumnApi: () => null,
};

export default SelectedForPlanning;
