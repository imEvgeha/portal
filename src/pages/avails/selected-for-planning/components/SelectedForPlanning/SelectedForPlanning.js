import React from 'react';
import PropTypes from 'prop-types';
import {compose} from 'redux';
import NexusGrid from '../../../../../ui/elements/nexus-grid/NexusGrid';
import ClickableCellRenderer from '../../../../../ui/elements/nexus-grid/elements/cell-renderer/clickable-cell-renderer/ClickableCellRenderer';
import withColumnsResizing from '../../../../../ui/elements/nexus-grid/hoc/withColumnsResizing';
import withInfiniteScrolling from '../../../../../ui/elements/nexus-grid/hoc/withInfiniteScrolling';
import withSideBar from '../../../../../ui/elements/nexus-grid/hoc/withSideBar';
import DOPService from '../../DOP-services';
import {COLUMN_MAPPINGS, SELECTED_FOR_PLANNING_TAB} from '../../constants';

const prepareSelectForPlanningData = async (sort, offset, limit) => {
    // Using object for easier parsing of getProjectAttributes response
    let data = {};

    // Fetch all active projects for the current user
    // N.B. offset + 1 because DOP API starts counting pages from 1
    //      while the rest of the APIs start from 0
    const [projectsList, headers]  = await DOPService.getUsersProjectsList(offset + 1, 1);
    const projectIds = [];

    // Extract project IDs for incomplete projects to display in SelectForPlanning table
    // TODO: Use projectsList when BE defect is fixed
    [{id: 'Project8680', status: 'CIN PROGRESS'}, {id: 'Project8679', status: 'CIN PROGRESS'}].forEach(({id, status}) => {
        if (!['COMPLETED', 'CANCELLED'].includes(status)) {
            projectIds.push(id);
            data[id] = {status};
        }
    });

    // Fetch project data and build rowData for ag-grid
    const projectAttributes = await DOPService.getProjectAttributes(projectIds);

    projectAttributes.forEach(({code, value, projectId}) => {
        // Filter out unwanted fields
        if (!['PROJECT_NAME'].includes(code)) {
            data[projectId][code] = value;
            // Adding projectId to be used for starting DOP project
            // in cellRenderer's onClick
            data[projectId]['projectId'] = projectId;
        }
    });

    // Convert object to an array
    data = Object.values(data);
    const total = headers.get('X-Total-Count') && data.length;

    return new Promise(res => {
        res({
            page: offset,
            size: limit,
            total,
            data,
        });
    });
};

const SelectedForPlanningTable = compose(
    withColumnsResizing(),
    withSideBar(),
    withInfiniteScrolling({fetchData: prepareSelectForPlanningData})
)(NexusGrid);

const SelectedForPlanning = ({activeTab}) => {
    const updatedColumnDefs = COLUMN_MAPPINGS.map(mapping => {
        // Attaching cellRenderer and action to status field
        // as it's used for starting DOP project
        if (mapping.field === 'status') {
            return {
                ...mapping,
                cellRenderer: 'clickableCellRenderer',
                cellRendererParams: {
                    onClick: DOPService.startProject,
                    keyToDisplay: 'status',
                },
            };
        }

        return mapping;
    });

    return (
        <SelectedForPlanningTable
            id="selectedForPlanningRepo"
            columnDefs={updatedColumnDefs}
            rowSelection="multiple"
            suppressRowClickSelection
            isGridHidden={activeTab !== SELECTED_FOR_PLANNING_TAB}
            frameworkComponents={{
                clickableCellRenderer: ClickableCellRenderer,
            }}
        />
    );
};

SelectedForPlanning.propTypes = {
    activeTab: PropTypes.string.isRequired,
};

export default SelectedForPlanning;
