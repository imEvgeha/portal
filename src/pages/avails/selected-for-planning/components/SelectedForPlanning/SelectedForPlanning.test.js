import React from 'react';
import {shallow} from 'enzyme';
import SelectedForPlanning from './SelectedForPlanning';

describe('SelectedForPlanning', () => {
    let wrapper = null;
    const props = {};
    beforeAll(() => {
        wrapper = shallow(<SelectedForPlanning {...props} />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
});