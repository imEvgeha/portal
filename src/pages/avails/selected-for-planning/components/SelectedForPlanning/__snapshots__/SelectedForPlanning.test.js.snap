// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedForPlanning should match snapshot 1`] = `
<Connect(ComposedComponent)
  columnDefs={
    Array [
      Object {
        "colId": "id",
        "field": "rightID",
        "headerName": "Right ID",
        "width": 120,
      },
      Object {
        "colId": "title",
        "field": "title",
        "headerName": "Title",
        "width": 120,
      },
      Object {
        "colId": "territory",
        "field": "territory",
        "headerName": "Planned Territory",
        "width": 150,
      },
      Object {
        "colId": "keywords",
        "field": "keywords",
        "headerName": "Keywords",
        "width": 150,
      },
      Object {
        "cellRenderer": "clickableCellRenderer",
        "cellRendererParams": Object {
          "keyToDisplay": "status",
          "onClick": [Function],
        },
        "colId": "status",
        "field": "status",
        "headerName": "DOP Status",
        "width": 180,
      },
      Object {
        "colId": "format",
        "field": "format",
        "headerName": "Format",
        "width": 120,
      },
      Object {
        "colId": "licensor",
        "field": "licensor",
        "headerName": "Licensor",
        "width": 120,
      },
      Object {
        "colId": "licensee",
        "field": "licensee",
        "headerName": "Licensee",
        "width": 120,
      },
      Object {
        "colId": "licenseType",
        "field": "licenseType",
        "headerName": "License type",
        "width": 120,
      },
      Object {
        "colId": "transactionType",
        "field": "transactionType",
        "headerName": "License type",
        "width": 120,
      },
    ]
  }
  frameworkComponents={
    Object {
      "clickableCellRenderer": [Function],
    }
  }
  id="selectedForPlanningRepo"
  isGridHidden={true}
  rowSelection="multiple"
  suppressRowClickSelection={true}
/>
`;
